pyro/poutine/handlers.py

```diff
@@ -69,359 +70,76 @@
+    :param fn: a stochastic function (callable containing Pyro primitive calls).
+    :param int rng_seed: rng seed.
     """
-    Given a stochastic function with some sample statements
-    and a dictionary of values at names,
-    set the return values of those sites equal to the values
-    as if they were hard-coded to those values
-    and introduce fresh sample sites with the same names
-    whose values do not propagate.
-
-    Composes freely with :func:`~pyro.poutine.handlers.condition`
-    to represent counterfactual distributions over potential outcomes.
-    See Single World Intervention Graphs [1] for additional details and theory.
-
-    Consider the following Pyro program:
-
-        >>> def model(x):
-        ...     s = pyro.param("s", torch.tensor(0.5))
-        ...     z = pyro.sample("z", dist.Normal(x, s))
-        ...     return z ** 2
+    def __init__(self, rng_seed):
+        assert isinstance(rng_seed, int)
+        self.rng_seed = rng_seed
+        super(SeedMessenger, self).__init__()
 
-    To intervene with a value for site `z`, we can write
+    def __enter__(self):
+        self.old_state = get_rng_state()
+        set_rng_seed(self.rng_seed)
 
-        >>> intervened_model = pyro.poutine.do(model, data={"z": torch.tensor(1.)})
+    def __exit__(self, type, value, traceback):
+        set_rng_state(self.old_state)
 
-    This is equivalent to replacing `z = pyro.sample("z", ...)` with
-    `z = torch.tensor(1.)`
-    and introducing a fresh sample site pyro.sample("z", ...) whose value is not used elsewhere.
 
-    References
+############################################
+# Begin primitive operations
+############################################
 
-    [1] `Single World Intervention Graphs: A Primer`,
-        Thomas Richardson, James Robins
 
-    :param fn: a stochastic function (callable containing Pyro primitive calls)
-    :param data: a ``dict`` mapping sample site names to interventions
-    :returns: stochastic function decorated with a :class:`~pyro.poutine.do_messenger.DoMessenger`
-    """
-    msngr = DoMessenger(data=data)
-    return msngr(fn) if fn is not None else msngr
+_msngrs = [
+    BlockMessenger,
+    BroadcastMessenger,
+    ConditionMessenger,
+    DoMessenger,
+    EnumMessenger,
+    EscapeMessenger,
+    InferConfigMessenger,
+    LiftMessenger,
+    MarkovMessenger,
+    MaskMessenger,
+    ReplayMessenger,
+    ScaleMessenger,
+    SeedMessenger,
+    TraceMessenger,
+    UnconditionMessenger,
+]
+
+
+def _make_handler(msngr_cls):
+
+    def handler(fn=None, *args, **kwargs):
+        if fn is not None and not callable(fn):
+            raise ValueError(
+                "{} is not callable, did you mean to pass it as a keyword arg?".format(fn))
+        msngr = msngr_cls(*args, **kwargs)
+        return msngr(fn) if fn is not None else msngr
+
+    return handler
+
+
+_re1 = re.compile('(.)([A-Z][a-z]+)')
+_re2 = re.compile('([a-z0-9])([A-Z])')
+for _msngr_cls in _msngrs:
```

Comment:
Here is the programmatic handler generation; compare to `pyro/distributions/torch.py`

Reply:
@neerajprad will this play nicely with serialization?  Do I need to use `functools.partial` here instead of a closure?

Reply:
This should work fine! These will be available from the module namespace, so pickle shouldn't have a problem.